src/data_store/expiration.clj

```diff
@@ -0,0 +1,31 @@
+(ns data-store.expiration
+  (:require [clojure.set :refer [union]]))
+
+(defn expired [store time]
+  (reduce
+   (fn [acc [k v]]
+     (if (and (:exp v) (>= (:exp v) time))
+       (conj acc k)
+       acc))
+   #{}
+   store))
+
+(defn has-exp [store]
+  (reduce
+   (fn [acc [k v]]
+     (if (contains? v :exp)
+       (into acc {k v})
+       acc))
+   {}
+   store))
+
+(defn expire-n [store time n]
+  (let [keys-expired
+        (loop [iter-keys-expired (expired (take n (has-exp store)) time)]
+          (if
+           (or (<= (count iter-keys-expired) (* 0.25 n)) (= (count iter-keys-expired) (count store)))
+            iter-keys-expired
+            (recur (union
+                    iter-keys-expired
+                    (expired (take n (has-exp (apply dissoc store iter-keys-expired))) time)))))]
+    (apply dissoc store keys-expired)))
```

Comment:
Runs the key expiration algorithm.

Gets `n` keys that have an expiration, and sees how many are expired.

If more than 25% are, it keeps getting `n` more.

The final `(apply dissoc` call removes the expired keys from the store.